src/upgrades/mongodb_upgrade.py

```diff
@@ -131,6 +194,147 @@ def _set_upgrade_status(self):
+            self.charm.mongodb_config, collection_name, write_value
+        )
+        self.clear_tmp_collection(self.charm.mongodb_config, collection_name)
+        return write_replicated
+
+    def clear_tmp_collection(
+        self, mongodb_config: MongoDBConfiguration, collection_name: str
+    ) -> None:
+        """Clears the temporary collection."""
+        with MongoDBConnection(mongodb_config) as mongod:
+            db = mongod.client["admin"]
+            db.drop_collection(collection_name)
+
+    @retry(
+        stop=stop_after_attempt(10),
+        wait=wait_fixed(1),
+        reraise=True,
+    )
+    def confirm_excepted_write_on_replica(
+        self,
+        host: str,
+        db_name: str,
+        collection: str,
+        expected_write_value: str,
+        secondary_config: MongoDBConfiguration,
+    ) -> bool:
+        """Returns True if the replica contains the expected write in the provided collection."""
+        secondary_config.hosts = {host}
+        with MongoDBConnection(secondary_config, direct=True) as direct_seconary:
+            db = direct_seconary.client[db_name]
+            test_collection = db[collection]
+            query = test_collection.find({}, {WRITE_KEY: 1})
+            if query[0][WRITE_KEY] != expected_write_value:
+                raise ClusterNotHealthyError
+
+    def get_random_write_and_collection(self) -> Tuple[str, str]:
+        """Returns a tuple for a random collection name and a unique write to add to it."""
+        choices = string.ascii_letters + string.digits
+        collection_name = "collection_" + "".join([secrets.choice(choices) for _ in range(32)])
+        write_value = "unique_write_" + "".join([secrets.choice(choices) for _ in range(16)])
+        return (collection_name, write_value)
+
+    def add_write_to_replica_set(
+        self, mongodb_config: MongoDBConfiguration, collection_name, write_value
+    ) -> None:
+        """Adds a the provided write to the admin database with the provided collection."""
+        with MongoDBConnection(mongodb_config) as mongod:
+            db = mongod.client["admin"]
+            test_collection = db[collection_name]
+            write = {WRITE_KEY: write_value}
+            test_collection.insert_one(write)
+
+    def is_write_on_secondaries(
+        self,
+        mongodb_config: MongoDBConfiguration,
+        collection_name,
+        expected_write_value,
+        db_name: str = "admin",
+    ):
+        """Returns true if the expected write."""
+        with MongoDBConnection(mongodb_config) as mongod:
+            primary_ip = mongod.primary()
+
+        replica_ips = mongodb_config.hosts
+        secondary_ips = replica_ips - set(primary_ip)
+        for secondary_ip in secondary_ips:
+            try:
+                self.confirm_excepted_write_on_replica(
+                    secondary_ip, db_name, collection_name, expected_write_value, mongodb_config
+                )
+            except RetryError:
```

Comment:
I believe this should be `except ClusterNotHealthyError` - given you're using `reraise=True`

Reply:
good catch, thank you!